### ISSUE TITLE
[QNX7.1] `std::fs::tests::symlink_hard_link` broken on main

### ISSUE THREAD

**Issue by japaric** [rust-lang/rust#129895]
Ferrocene CI has detected that this test was broken by rust-lang/rust#127897 . Specifically, by the change in `library/std/src/sys/pal/unix/fs.rs`, shown below:

``` diff
diff --git a/library/std/src/sys/pal/unix/fs.rs b/library/std/src/sys/pal/unix/fs.rs
index 7fa147c9754..fc9d7e98883 100644
--- a/library/std/src/sys/pal/unix/fs.rs
+++ b/library/std/src/sys/pal/unix/fs.rs
@@ -1717,7 +1717,7 @@ pub fn link(original: &Path, link: &Path) -> io::Result<()> {
     run_path_with_cstr(original, &|original| {
         run_path_with_cstr(link, &|link| {
             cfg_if::cfg_if! {
-                if #[cfg(any(target_os = "vxworks", target_os = "redox", target_os = "android", target_os = "espidf", target_os = "horizon", target_os = "vita", target_os = "nto"))] {
+                if #[cfg(any(target_os = "vxworks", target_os = "redox", target_os = "android", target_os = "espidf", target_os = "horizon", target_os = "vita"))] {
                     // VxWorks, Redox and ESP-IDF lack `linkat`, so use `link` instead. POSIX leaves
                     // it implementation-defined whether `link` follows symlinks, so rely on the
                     // `symlink_hard_link` test in library/std/src/fs/tests.rs to check the behavior.
```

Test output:

``` console
---- fs::tests::symlink_hard_link stdout ----
thread 'fs::tests::symlink_hard_link' panicked at std/src/fs/tests.rs:1473:5:
assertion failed: check!(fs::symlink_metadata(tmpdir.join("hard_link"))).file_type().is_symlink()

```

Reverting that single line diff fixes the test for the QNX7.1 targets, e.g. `arch64-unknown-nto-qnx710` and `x86_64-pc-nto-qnx710`

@nyurik can the change be reverted or does QNX7.0 need to use `link`, instead of `linkat`, here? from looking at libc, it appears that both `link` and `linkat` are available on QNX7.0 so reverting the change should at least not cause compilation or linking errors. if the case is the latter, then we should use `cfg(target_env = "nto70")` in addition to `cfg(target_os = "nto")`.

**Comment by nyurik** [comment]
@japaric thanks for the report! I will need to re-test it for QNX 7.0, to see if it is a 7.0 requirement (i don't recall why this was required initially, but clearly I ran into some issues with it). I also plan to do a similar automation to ensure 7.0 builds run without problems on our hardware. Is there a list of tests you see failing on 7.1 that are OK to ignore?  I see a list [here](https://doc.rust-lang.org/rustc/platform-support/nto-qnx.html#running-the-rust-test-suite) - is this the most relevant one? Thx!

**Comment by japaric** [comment]
we do not ignore any single unit test at the moment and we run library (e.g. libstd) tests as well as (cross) compilation tests using `TEST_DEVICE_ADDR`. we don't pass `--exclude` to `x.py` but instead pass the list of test suites as arguments.

we currently run these test suites:

``` console
$ # run these from the root of the ferrocene/ferrocene repo

$ ferrocene/ci/split-tasks.py test:library
library/alloc library/core library/test

$ ferrocene/ci/split-tasks.py test:library-std
library/std

$ ferrocene/ci/split-tasks.py qnx:compiletest-no-only-hosts
tests/assembly tests/codegen tests/codegen-units tests/coverage tests/crashes tests/debuginfo tests/incremental tests/mir-opt tests/run-make tests/run-pass-valgrind tests/ui
```

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23129895.20.5BQNX7.2E1.5D.20.60std.3A.3Afs.3A.3Atests.3A.3Asymlink_hard_link.60.20broken.E2.80.A6)).

@rustbot label -I-prioritize +P-low

**Comment by nyurik** [comment]
fixed in #130248
